gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java

```
@@ -334,78 +327,136 @@ private static <T> ObjectConstructor<T> newDefaultImplementationConstructor(
-          }
-        };
-      } else {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new ArrayList<>();
-          }
-        };
-      }
+      @SuppressWarnings("unchecked")
+      ObjectConstructor<T> constructor = (ObjectConstructor<T>) newCollectionConstructor(rawType);
+      return constructor;
     }
 
     if (Map.class.isAssignableFrom(rawType)) {
-      if (ConcurrentNavigableMap.class.isAssignableFrom(rawType)) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new ConcurrentSkipListMap<>();
-          }
-        };
-      } else if (ConcurrentMap.class.isAssignableFrom(rawType)) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new ConcurrentHashMap<>();
-          }
-        };
-      } else if (SortedMap.class.isAssignableFrom(rawType)) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new TreeMap<>();
-          }
-        };
-      } else if (type instanceof ParameterizedType
-          && !String.class.isAssignableFrom(
-              TypeToken.get(((ParameterizedType) type).getActualTypeArguments()[0]).getRawType())) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new LinkedHashMap<>();
-          }
-        };
-      } else {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new LinkedTreeMap<>();
-          }
-        };
-      }
+      @SuppressWarnings("unchecked")
+      ObjectConstructor<T> constructor = (ObjectConstructor<T>) newMapConstructor(type, rawType);
+      return constructor;
+    }
+
+    // Unsupported type; try other means of creating constructor
+    return null;
```

In https://github.com/google/gson/issues/419#issuecomment-83705342 the concern was mentioned that the fallback will be `Unsafe` which might leave the Collection or Map implementation in a broken state because fields have not been initialized. And the built-in Collection and Map adapters might then encounter confusing exceptions when trying to add elements.

Not sure if we should throw an exception then in case of Collection or Map? Or we could maybe add an `allowUnsafe` parameter here (which is considered in addition to `GsonBuilder#disableJdkUnsafe`), which the built-in Collection and Map adapters can set to `false` because calling `add` and `put` will likely fail afterwards.

---

@eamonnmcmanus, do you think this is something to consider?

---

I'm not sure I understand the concern here. If we reach this point then `rawType` is neither a `Collection` nor a `Map`. Could you explain what sort of type you might consider throwing an exception for here?

---

Seems like there is a [bug with GitHub](https://github.com/orgs/community/discussions/150115#discussioncomment-12294289) and it is not showing the comment I was referring to above; here is an [archived version](https://web.archive.org/web/20200917230215/https://github.com/google/gson/issues/419#issuecomment-83705342):

> I'm anxious about using UnsafeAllocator on map and collection types. Types like HashMap expect their constructors to run (and initialize state) but they won't necessarily get run if we rely on UnsafeAllocator. (We'd be relying on the presence of a no-args constructor).

When users define custom `Collection` or `Map` implementations without no-args constructor, then `Unsafe` would be used to create the instance. But because the Collection and Map adapter calls Collection and Map methods during deserialization, that most likely leads to confusing exceptions because Unsafe did not properly initialize the instances.

I have pushed a commit now which prevents Unsafe usage in this case; I hope that is fine. If you don't think it should be included please let me know and I revert it again.

---

OK, I think I understand the concern now. It is not about this statement particularly, but the fact that the earlier `return` statements can return null if `newCollectionConstructor` or `newMapConstructor` does, and then we'll fall back to the `Unsafe` path.

We don't see this issue with Guava's `ImmutableList` etc because the public types are abstract, so `Unsafe` obviously can't instantiate them. I generally hate the nonsense with `Unsafe` allocation so I'm all for changes that allow less of it. However I'll want to run this latest version past Google's tests again to see if there are further unforeseen consequences.